realm/realm-library/src/main/java/io/realm/BaseRealm.java

```diff
@@ -152,14 +161,25 @@ public boolean isInTransaction() {
      */
     public void addChangeListener(RealmChangeListener listener) {
         checkIfValid();
-        for (WeakReference<RealmChangeListener> ref : changeListeners) {
+        changeListeners.addIfAbsent(listener);
+    }
+
+    /**
+     * For internal use only.
+     * Sometimes we don't know when to unregister listeners (ex: {@link RealmBaseAdapter}), using
+     * a WeakReference the listener.
+     *
+     * @param listener the change listener.
+     */
+    void addChangeListenerAsWeakReference(RealmChangeListener listener) {
```

Comment:
You're already making `getRealm` public in https://github.com/realm/realm-java/pull/1093 
this should work + `addChangeListenerAsWeakReference` is not Android specific.
It's just a behaviour we don't want to expose as public API.